Chrome system tests: ensure the page is fully loaded (including all iframes) before testing the page

michaelDCurran opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 
-->

### Link to issue number:
None.

### Summary of the issue:
Recently some of our system tests for Chrome started breaking. Specifically the tests that include an iframe.
It looks as if the test logic is starting to run before the content in the iframe is fully loaded.

### Description of how this pull request fixes the issue:
Place an initial line at the top of the test page, before the start marker. NVDA will initially land on this line as soon as the document gets focus, but possibly before iframes have fully loaded.
The page now has an onload event which forces focus to the start marker. This onload event only fires once all content (including iframes) in the document are fully loaded.
When the start marker finally gets focus, NVDA will detect this and start running the test logic.

### Testing strategy:
Successfully ran all system tests locally.

### Known issues with pull request:
None known.

### Change log entry:
None needed.

### Code Review Checklist:

This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review and confirm you have considered the following items.
Mark items you have considered by checking them.
You can do this when editing the Pull request description with an x: `[ ]` becomes `[x]`.
You can also check the checkboxes after the PR is created.

- [ ] Pull Request description is up to date.
- [ ] Unit tests.
- [ ] System (end to end) tests.
- [ ] Manual tests.
- [ ] User Documentation.
- [ ] Change log entry.
- [ ] Context sensitive help for GUI changes.

michaelDCurran committed:
Chrome system tests: add a load status line after the start marker, and wait for it to become "Test page load complete" before continuing with the test. This is set by the boy's onload event.

michaelDCurran committed:
Set the NVDA startup complete variable from a queued lambda to ensure that NvDA's core loop has fully started and the initial focus event has occured.

michaelDCurran committed:
Linting

michaelDCurran committed:
Add comment.

feerrenrut commented on tests/system/libraries/SystemTestSpy/speechSpyGlobalPlugin.py:
```diff
@@ -68,7 +68,9 @@ def _registerWithExtensionPoints(self):
 	# callbacks for extension points
 	def _onNvdaStartupComplete(self):
-		self._isNvdaStartupComplete = True
+		# Queue the setting of the completion variable,
+		# To ensure that NvDA's core loop has started running, and it has processed initial focus.
+		queueHandler.queueFunction(queueHandler.eventQueue, lambda: setattr(self, '_isNvdaStartupComplete', True))
```
Perhaps the code at `source/core.py:564: postNvdaStartup.notify()` should be queued instead? If we are saying the loop must have started before NVDA's startup is complete, then the `postNvdaStartup` action is incorrect.

michaelDCurran replied:
I think for system tests it makes sense to ensure the loop has started and initial focus is processed. But for other uses I'm not sure. I'm more inclined just to change the system tests for now, rather than change the behaviour of that action, which other 3rd party code *could* be using.

feerrenrut reviewed (changes requested):
Overall looks good, couple of minor suggestions.

michaelDCurran committed:
remove unneeded space.

michaelDCurran committed:
Core: queue the PostNVDAStartup notification so that it is actually executed from within the core loop once running, and after any initial focus has been reported.
this is rather than systemTestSpy queuing the setting of its startupComplete variable.

feerrenrut reviewed (approved):
Looks good to me, assuming the tests pass.
